### PR TITLE
cpu/saml21: add DFLL support

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description

This PR adds DFLL support (48MHz) for SAML21 to be used to feed CPU clock.
I ran several tests like xtimer, I2C, SPI and they all succeeded.

### Testing procedure
Ensure `#define DFLL_CLK1 `and `#define CLOCK_CORECLOCK 48000000u` are correctly set in `periph_conf.h`
run all the tests you want !

### Issues/PRs references
none but needed for a futur USB support :)

### CODE REVIEW

**Comment by dylad** [comment]
@aabadie  May I ask you to have a look at this (after the current release cycle) please ?
I just update it to add SAMR30-XPRO board.

**Comment by dylad** [comment]
Rebased this PR, but we must now wait for #11502 to be merged first.

**Comment by dylad** [comment]
with #11502 merged, this PR should be ready for review.

**Comment by fjmolinas** [comment]
@dylad this one needs rebasing!

**Comment by fjmolinas** [comment]
ping @dylad

**Comment by dylad** [comment]
rebased and reworked.
Update SAMR34 periph_conf.h
Switch saml21-based boards to 48000000 MHz by default
I also rework the dfll selection a bit to simply it for users.

File: cpu/saml21/cpu.c [hunk 1]
```diff
@@ -106,10 +153,18 @@ void cpu_init(void)
...
     _osc32k_setup();
     _xosc32k_setup();
+    _dfll_setup();
 
     /* Setup GCLK generators */
+#if (CLOCK_CORECLOCK == 16000000U)
     _gclk_setup(0, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_OSC16M);
-
+#elif (CLOCK_CORECLOCK == 48000000U)
+     _gclk_setup(0, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M);
```
Comment (aabadie) [inline cpu/saml21/cpu.c]:
```suggestion
    _gclk_setup(0, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M);
```

Comment (dylad) [reply cpu/saml21/cpu.c]: Fixed.

File: cpu/saml21/cpu.c [hunk 2]
```diff
@@ -64,6 +64,53 @@ static void _xosc32k_setup(void)
 #endif
 }
 
+static void _dfll_setup(void)
+{
+#if (CLOCK_CORECLOCK == 48000000U) || defined (MODULE_PERIPH_USBDEV)
```
Comment (aabadie) [inline cpu/saml21/cpu.c]: Why not put the function definition inside the macros:
```c
#if (CLOCK_CORECLOCK == 48000000U) || defined (MODULE_PERIPH_USBDEV)
static void _dfll_setup(void)
{
...
}
#endif
```

and later call the function only when core clock is 48MHz and usbdev module is used.

**Comment by aabadie** [comment]
I tested on sam34-xpro (only saml21 based board I have with me ATM): basic xtimer tests, periph_timer test work. `tests/periph_rtc` works with this PR (but doesn't in master).
Only problem: `tests/periph_rtt` doesn't work anymore with this PR (it works on master).

I'll see if I can free some bandwidth today to run the `compile_and_test_for_board.py` script on samr34-xpro. @dylad, it would be great if you could report the same for saml21-xpro.

File: cpu/saml21/cpu.c (hunk 2 above)
Comment (dylad) [reply cpu/saml21/cpu.c]: This would force me to add another guard when calling _dfll_setup and thus decrease readability. Should I change it anyways ?

Comment (aabadie) [reply cpu/saml21/cpu.c]: > add another guard when calling _dfll_setup

I know ;)

> Should I change it anyways ?

It depends: if it saves some ROM when using a 16MHz coreclock then yes.

Comment (dylad) [reply cpu/saml21/cpu.c]: > It depends: if it saves some ROM when using a 16MHz coreclock then yes.

I tested this specific point, there is no RAM/ROM changes if we keep it this way. Hopefully :)

**Comment by dylad** [comment]
Force-pushed to rebase against current master to include recent changes.
I did NOT test these changes yet.

> See what is done for hifive1b for example.
For now I think it's ok and it could be done in follow-up PRs.

I'll take a deep look at it as I really want to rework the cpu.c file for saml21 right after this PR.
For now, I'd like to enable USB on SAML21 for users (with this PR + #12132)

**Comment by dylad** [comment]
`dist/tools/compile_and_test_for_board/compile_and_test_for_board.py`
reports
`- [tests/periph_rtt](tests/periph_rtt/test.failed)`

I'll investigate further.

**Comment by benpicco** [comment]
btw: I also have a `saml21-xpro` now, so I can give this a try.

**Comment by dylad** [comment]
I still have no idea what's wrong with periph_rtt...
Looks like the reset at the beginning of the init is the culprit but this is a bit crazy...

**Comment by dylad** [comment]
I'm facing the weirdest ever bug on this MCU...
Resetting RTC with RTC->MODE0.bit.SWRST seems to kill XOSC32K and thus the DFLL which is feed by 32kHz clock. I dunno how to handle this...
@aabadie regarding your previous comment. It seems NOT possible to use the DFLL48M source with a clock different than 48MHz.

**Comment by benpicco** [comment]
What if we just don't reset the RTT like we do with RTC?

**Comment by dylad** [comment]
> What if we just don't reset the RTT like we do with RTC?

Then it works flawlessly.
If we don't reset the RTT value during the init, the count register will not be clear after wakeup from standby or reset. I don't think this is a desirable behavior.

**Comment by benpicco** [comment]
> If we don't reset the RTT value during the init, the count register will not be clear after wakeup from ~~standby or~~ reset. I don't think this is a desirable behavior.

Hm, I don't think it would be so bad.
Users of `rtt_set_alarm()` will need to call `rtt_get_counter()` anyway and typically the RTT will wrap after 1½ days, so it's not useful as an uptime counter or such.

AFAIR Linux initializes timers to a couple of minutes before they overflow, so bugs are caught early.
I don't think the absolute value of a timer is significant.

**Comment by dylad** [comment]
Removing it is not a big deal but this should belongs to another PR as it affects all SAM0. Moreover, this bug is just too weird to be left alive ! It drives me crazy...

**Comment by dylad** [comment]
This is killing me but I want to merge this PR...
@aabadie could you run `compile_and_test_for_board` on a SAMD21-based MCU to ensure I didn't break stuff please ? It is time to move forward.

**Comment by aabadie** [comment]
I tested on samr34-xpro (saml21) and samr21-xpro and this PR breaks `periph_rtt` on both boards, compared to master.
To verify this, flash and test `tests/periph_rtc`, then `tests/periph_rtt`. On master, both works, with this PR, the latter doesn't. The call to the reset function should be kept I think.

**Comment by dylad** [comment]
> On master, both works, with this PR, the latter doesn't. The call to the reset function should be kept I think.

I must be cursed or something...

File: tests/periph_rtt/main.c [hunk 3]
```diff
@@ -48,7 +48,7 @@ void cb(void *arg)
 
 int main(void)
 {
-    test_utils_interactive_sync();
+   //test_utils_interactive_sync();
```
Comment (gschorcht) [inline tests/periph_rtt/main.c]: Shouldn't be accidentally merged.

Comment (dylad) [reply tests/periph_rtt/main.c]: Don't worry, It was pushed on purpose. I also remove the ready for build label.
This commit will be dropped as soon as the crazy bug will be fixed. But for now, I'm still hunting.

**Comment by dylad** [comment]
@aabadie could you retry the current branch on your hardware next week please ?
I just tried on iotlab and it runs fine...

**Comment by jue89** [comment]
I tested the `samr30-xpro` board. `periph_rtc` works. `periph_rtt` doesn't. On master everything is fine. I also tried to rebase this PR on master. But this doesn't improve the situation :/

**Comment by dylad** [comment]
> I tested the samr30-xpro board. periph_rtc works. periph_rtt doesn't. On master everything is fine. I also tried to rebase this PR on master. But this doesn't improve the situation :/

I'm still hunting this issue. But if you can reproduce it it means my hardware isn't that crazy ;)
It _looks_ like we have some stability issue with XOSC32K when sourcing DFLL + RTC...

**Comment by dylad** [comment]
Rebased and squashed.
The RTT issue has been solved. This PR is ready for its final test and SAML21 is almost ready for USB support.

File: cpu/saml21/cpu.c [hunk 4]
```diff
@@ -125,14 +169,34 @@ void cpu_init(void)
...
+        if(CLOCK_CORECLOCK / (1 << i) <= 6000000) {
+            MCLK->BUPDIV.reg = 1 << i;
+        }
+    }
+
+    /* Setup GCLK generators */
+#if (CLOCK_CORECLOCK == 16000000U)
+    _gclk_setup(SAM0_GCLK_MAIN, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_OSC16M);
+#elif (CLOCK_CORECLOCK == 48000000U)
+     _gclk_setup(SAM0_GCLK_MAIN, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M);
```
Comment (aabadie) [inline cpu/saml21/cpu.c]:
```suggestion
    _gclk_setup(SAM0_GCLK_MAIN, GCLK_GENCTRL_GENEN | GCLK_GENCTRL_SRC_DFLL48M);
```

File: cpu/saml21/cpu.c [hunk 5]
```diff
@@ -75,13 +75,57 @@ uint32_t sam0_gclk_freq(uint8_t id)
...
+    while (!(OSCCTRL->STATUS.reg & OSCCTRL_STATUS_DFLLRDY)) {}
+    /* Write full configuration to DFLL control register */
+    OSCCTRL->DFLLCTRL.reg =  OSCCTRL_DFLLCTRL_WAITLOCK |
+                             OSCCTRL_DFLLCTRL_MODE |
+                             OSCCTRL_DFLLCTRL_CCDIS |
+                             OSCCTRL_DFLLCTRL_BPLCKC |
+                             OSCCTRL_DFLLCTRL_ENABLE;
+
+    /* Ensure COARSE and FINE are locked */
+    while((!(OSCCTRL->STATUS.bit.DFLLLCKC)) && (!(OSCCTRL->STATUS.bit.DFLLLCKF))) {}
```
Comment (aabadie) [inline cpu/saml21/cpu.c]:
```suggestion
    while ((!(OSCCTRL->STATUS.bit.DFLLLCKC)) && (!(OSCCTRL->STATUS.bit.DFLLLCKF))) {}
```

File: cpu/saml21/cpu.c (hunk 4 above)
Comment (dylad) [reply cpu/saml21/cpu.c]: Fixed.

File: cpu/saml21/cpu.c (hunk 5 above)
Comment (dylad) [reply cpu/saml21/cpu.c]: Fixed.

**Comment by aabadie** [comment]
Unfortunately, I still have problems with `tests/periph_rtt`, other applications work well:

<details><summary>samr34-xpro</summary>

```
$ make BOARD=samr34-xpro -C tests/periph_rtt flash test --no-print-directory 
Building application "tests_periph_rtt" for "samr34-xpro" with MCU "saml21".

"make" -C /work/riot/RIOT-review/boards/samr34-xpro
"make" -C /work/riot/RIOT-review/core
"make" -C /work/riot/RIOT-review/cpu/saml21
"make" -C /work/riot/RIOT-review/cpu/cortexm_common
"make" -C /work/riot/RIOT-review/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT-review/cpu/sam0_common
"make" -C /work/riot/RIOT-review/cpu/sam0_common/periph
"make" -C /work/riot/RIOT-review/cpu/saml21/periph
"make" -C /work/riot/RIOT-review/drivers
"make" -C /work/riot/RIOT-review/drivers/periph_common
"make" -C /work/riot/RIOT-review/sys
"make" -C /work/riot/RIOT-review/sys/auto_init
"make" -C /work/riot/RIOT-review/sys/isrpipe
"make" -C /work/riot/RIOT-review/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT-review/sys/pm_layered
"make" -C /work/riot/RIOT-review/sys/stdio_uart
"make" -C /work/riot/RIOT-review/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT-review/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  10340	    136	   2636	  13112	   3338	/work/riot/RIOT-review/tests/periph_rtt/bin/samr34-xpro/tests_periph_rtt.elf
/work/riot/RIOT-review/dist/tools/edbg/edbg  --target samr34 --verbose --file /work/riot/RIOT-review/tests/periph_rtt/bin/samr34-xpro/tests_periph_rtt.bin --verify || /work/riot/RIOT-review/dist/tools/edbg/edbg  --target samr34 --verbose --file /work/riot/RIOT-review/tests/periph_rtt/bin/samr34-xpro/tests_periph_rtt.bin --verify --program
Debugger: ATMEL EDBG CMSIS-DAP ATML3167041800000808 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM R34J18B (Rev C)
Verification............................................ done.
r
/work/riot/RIOT-review/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2020.04-devel-1635-ge1aeb-review_dfll)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
Timeout in expect script at "child.expect(r'RTT now: \d+')" (tests/periph_rtt/tests/01-run.py:22)
```

</details>

<details><summary>samr30-xpro (on iotlab)</summary>

```
$ make BOARD=samr30-xpro -C tests/periph_rtt flash test --no-print-directory IOTLAB_NODE=auto-ssh
Building application "tests_periph_rtt" for "samr30-xpro" with MCU "saml21".

"make" -C /work/riot/RIOT-review/boards/samr30-xpro
"make" -C /work/riot/RIOT-review/core
"make" -C /work/riot/RIOT-review/cpu/saml21
"make" -C /work/riot/RIOT-review/cpu/cortexm_common
"make" -C /work/riot/RIOT-review/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT-review/cpu/sam0_common
"make" -C /work/riot/RIOT-review/cpu/sam0_common/periph
"make" -C /work/riot/RIOT-review/cpu/saml21/periph
"make" -C /work/riot/RIOT-review/drivers
"make" -C /work/riot/RIOT-review/drivers/periph_common
"make" -C /work/riot/RIOT-review/sys
"make" -C /work/riot/RIOT-review/sys/auto_init
"make" -C /work/riot/RIOT-review/sys/isrpipe
"make" -C /work/riot/RIOT-review/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT-review/sys/pm_layered
"make" -C /work/riot/RIOT-review/sys/stdio_uart
"make" -C /work/riot/RIOT-review/sys/test_utils/interactive_sync
"make" -C /work/riot/RIOT-review/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  10436	    136	   2636	  13208	   3398	/work/riot/RIOT-review/tests/periph_rtt/bin/samr30-xpro/tests_periph_rtt.elf
iotlab-node --jmespath='keys(@)[0]' --format='int'  --list saclay,samr30,1 --flash /work/riot/RIOT-review/tests/periph_rtt/bin/samr30-xpro/tests_periph_rtt.bin | grep 0
0
r
ssh -t abadie@saclay.iot-lab.info 'socat - tcp:samr30-1.saclay.iot-lab.info:20000'
READY
s
START
main(): This is RIOT! (Version: 2020.04-devel-1635-ge1aeb-review_dfll)

RIOT RTT low-level driver test
This test will display 'Hello' every 5 seconds

Initializing the RTT driver
Timeout in expect script at "child.expect(r'RTT now: \d+')" (tests/periph_rtt/tests/01-run.py:22)

```

</details>

File: cpu/saml21/cpu.c [hunk 6]
```diff
@@ -188,16 +196,17 @@ void cpu_init(void)
 
     /* Ensure APB Backup domain clock is within the 6MHZ limit, BUPDIV value
        must be a power of 2 and between 1(2^0) and 128(2^7) */
-        for (unsigned i=0; i<8; i++) {
+    for (unsigned i=0; i<8; i++) {
```
Comment (aabadie) [inline cpu/saml21/cpu.c]:
```suggestion
    for (unsigned i = 0; i < 8; i++) {
```

**Comment by aabadie** [comment]
Works now for me. Please squash!

File: cpu/saml21/cpu.c (hunk 6 above)
Comment (dylad) [reply cpu/saml21/cpu.c]: Done.

**Comment by dylad** [comment]
Squashed.

**Comment by dylad** [comment]
@gschorcht Your comment was addressed. Do you mind to dismiss your change request ?

**Comment by benpicco** [comment]
Murdock is being wonky…

```
	'make doc' exited with non-zero code (2)
	make: Entering directory '/tmp/dwq.0.9668434009531381/d51fbb33d15d484f561bcd74fe9ab34b'
	"make" -BC doc/doxygen
	make[1]: Entering directory '/tmp/dwq.0.9668434009531381/d51fbb33d15d484f561bcd74fe9ab34b/doc/doxygen'
	( cat riot.doxyfile ; echo "GENERATE_HTML = yes" ) | doxygen -
	sh: 1: latex: not found
	error: Problems running latex. Check your installation or look for typos in _formulas.tex and check _formulas.log!
	sh: 1: dvips: not found
	error: Problems running dvips. Check your installation!
	Killed
	Makefile:18: recipe for target 'html' failed
	make[1]: *** [html] Error 137
	make[1]: Leaving directory '/tmp/dwq.0.9668434009531381/d51fbb33d15d484f561bcd74fe9ab34b/doc/doxygen'
	Makefile:10: recipe for target 'doc' failed
	make: *** [doc] Error 2
	make: Leaving directory '/tmp/dwq.0.9668434009531381/d51fbb33d15d484f561bcd74fe9ab34b'
```

**Comment by aabadie** [comment]
ping @gschorcht !

**Comment by aabadie** [comment]
@gschorcht's change request is obviously fixed so there's no harm to dismiss it. Let's move forward!

**Comment by dylad** [comment]
Thanks a lot for testing and merging @aabadie !

**Pull request merged**